gym_xiangqi/envs/xiangqi_env.py

```diff
@@ -21,5 +178,278 @@ def render(self, mode='human'):
+
+        Parameters:
+            piece_id (int): a piece ID integer
+            start (tuple(int)): (row, col) start coordinate
+            end (tuple(int)): (row, col) end coordinate
+        return:
+            Index within the self.possible_actions
+        """
+        piece_id_val = (piece_id-1) * pow(self.total_pos, 2)
+        start_val = (start[0] * 9 + start[1]) * self.total_pos
+        end_val = end[0] * 9 + end[1]
+        return piece_id_val + start_val + end_val
+
+    def action_space_to_move(self, action):
+        """
+        This is exact opposite of move_to_action_space() method.
+        With index value, we can convert this back to piece id,
+        start position and end position values.
+
+        Parameters:
+            action (int): index value within action space
+        return:
+            piece ID, start coordinate, end coordinate
+        """
+        piece_id, r = divmod(action, pow(self.total_pos, 2))
+        start_val, end_val = divmod(r, self.total_pos)
+        start = [0, 0]
+        end = [0, 0]
+        start[0], start[1] = divmod(start_val, 9)
+        end[0], end[1] = divmod(end_val, 9)
+        return piece_id+1, start, end
+
     def get_possible_actions(self):
-        pass
+        """
+        Searches all valid actions each piece can perform
+        """
+        self.get_general_actions()
+        self.get_advisor_actions()
+        self.get_elephant_actions()
+        self.get_horse_actions()
+        self.get_chariot_actions()
+        self.get_cannon_actions()
+        self.get_soldier_actions()
+
+    def check_action(self, piece, orig_pos, cur_pos, repeat, offset, i):
+        """
+        This is general searching procedure. Given the following parameters,
+        repeatedly search in the same direction until either end of the board
+        or another piece is blocking.
+
+        Parameters:
+            piece (int): piece ID
+            orig_pos (tuple(int)): original position coordinate
+            cur_pos (tuple(int)): current position in evaluation
+            repeat (int): number of repetitions
+            offset (tuple(int)): direction offset
+            i (int): current iteration number
+        return:
+            Number of times repeated; This is used to find out the farthest
+            possible position.
+        """
+        r = cur_pos[0]
+        c = cur_pos[1]
+
+        for i in range(repeat):
+            row_bound = 0 <= r < self.shape[0]
+            col_bound = 0 <= c < self.shape[1]
+
+            if not row_bound or not col_bound:
+                return i
+
+            if self.state[r][c] > 0:
+                break
+
+            action_idx = self.move_to_action_space(piece, orig_pos, (r, c))
+            self.possible_actions[action_idx] = 1
+
+            if self.state[r][c] != 0:
+                break
+
+            r += offset[0]
+            c += offset[1]
+
+        return i + 1
+
+    def get_general_actions(self):
```

Comment:
Is it possible to move each of the piece's action into their own class and pass the game state to the pieces instead? And maybe put the shared method in the parent class. Feels like that is the more suitable place to have them.

Reply:
If it's too difficult, I am okay also.

Reply:
change made based on this